### PR TITLE
executor: get schema and outname of view once in the dataForColumnsInTable

### LINKED ISSUE #34657: DATA RACE in the dataForColumnsInTable
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

```
==================
WARNING: DATA RACE
Write at 0x00c005fd0d28 by goroutine 162:
  github.com/pingcap/tidb/executor.tryFillViewColumnType.func1()
      /go/tidb/executor/show.go:1913 +0x645
  github.com/pingcap/tidb/executor.runWithSystemSession()
      /go/tidb/executor/show.go:1940 +0x334
  github.com/pingcap/tidb/executor.tryFillViewColumnType()
      /go/tidb/executor/show.go:1904 +0x16d
  github.com/pingcap/tidb/executor.(*hugeMemTableRetriever).dataForColumnsInTable()
      /go/tidb/executor/infoschema_reader.go:680 +0x159
  github.com/pingcap/tidb/executor.(*hugeMemTableRetriever).setDataForColumns()
      /go/tidb/executor/infoschema_reader.go:669 +0x4a4
  github.com/pingcap/tidb/executor.(*hugeMemTableRetriever).retrieve()
      /go/tidb/executor/infoschema_reader.go:2614 +0x314
  github.com/pingcap/tidb/executor.(*MemTableReaderExec).Next()
      /go/tidb/executor/memtable_reader.go:118 +0x482
  github.com/pingcap/tidb/executor.Next()
      /go/tidb/executor/executor.go:319 +0x645
  github.com/pingcap/tidb/executor.(*SelectionExec).Next()
      /go/tidb/executor/executor.go:1411 +0x278
  github.com/pingcap/tidb/executor.Next()
      /go/tidb/executor/executor.go:319 +0x645
  github.com/pingcap/tidb/executor.(*HashJoinExec).fetchBuildSideRows()
      /go/tidb/executor/join.go:282 +0x43e
  github.com/pingcap/tidb/executor.(*HashJoinExec).fetchAndBuildHashTable.func2()
      /go/tidb/executor/join.go:731 +0xf9
  github.com/pingcap/tidb/util.WithRecovery()
      /go/tidb/util/misc.go:90 +0x6d
  github.com/pingcap/tidb/executor.(*HashJoinExec).fetchAndBuildHashTable.func4()
      /go/tidb/executor/join.go:728 +0x47
Previous read at 0x00c005fd0d28 by goroutine 64:
  github.com/pingcap/tidb/parser/types.(*FieldType).GetType()
      /go/tidb/parser/types/field_type.go:63 +0x6e6
  github.com/pingcap/tidb/parser/model.(*ColumnInfo).GetType()
      /go/tidb/parser/model/model.go:135 +0x6ca
  github.com/pingcap/tidb/executor.tryFillViewColumnType.func1()
      /go/tidb/executor/show.go:1915 +0x6b9
  github.com/pingcap/tidb/executor.runWithSystemSession()
      /go/tidb/executor/show.go:1940 +0x334
  github.com/pingcap/tidb/executor.tryFillViewColumnType()
      /go/tidb/executor/show.go:1904 +0x16d
  github.com/pingcap/tidb/executor.(*hugeMemTableRetriever).dataForColumnsInTable()
      /go/tidb/executor/infoschema_reader.go:680 +0x159
  github.com/pingcap/tidb/executor.(*hugeMemTableRetriever).setDataForColumns()
      /go/tidb/executor/infoschema_reader.go:669 +0x4a4
  github.com/pingcap/tidb/executor.(*hugeMemTableRetriever).retrieve()
      /go/tidb/executor/infoschema_reader.go:2614 +0x314
  github.com/pingcap/tidb/executor.(*MemTableReaderExec).Next()
      /go/tidb/executor/memtable_reader.go:118 +0x482
  github.com/pingcap/tidb/executor.Next()
      /go/tidb/executor/executor.go:319 +0x645
  github.com/pingcap/tidb/executor.(*SelectionExec).Next()
      /go/tidb/executor/executor.go:1411 +0x278
  github.com/pingcap/tidb/executor.Next()
      /go/tidb/executor/executor.go:319 +0x645
  github.com/pingcap/tidb/executor.(*HashJoinExec).fetchBuildSideRows()
      /go/tidb/executor/join.go:282 +0x43e
  github.com/pingcap/tidb/executor.(*HashJoinExec).fetchAndBuildHashTable.func2()
      /go/tidb/executor/join.go:731 +0xf9
  github.com/pingcap/tidb/util.WithRecovery()
      /go/tidb/util/misc.go:90 +0x6d
  github.com/pingcap/tidb/executor.(*HashJoinExec).fetchAndBuildHashTable.func4()
      /go/tidb/executor/join.go:728 +0x47
Goroutine 162 (running) created at:
  github.com/pingcap/tidb/executor.(*HashJoinExec).fetchAndBuildHashTable()
      /go/tidb/executor/join.go:728 +0x405
  github.com/pingcap/tidb/executor.(*HashJoinExec).Next.func1()
      /go/tidb/executor/join.go:687 +0xc8
  github.com/pingcap/tidb/util.WithRecovery()
      /go/tidb/util/misc.go:90 +0x6d
  github.com/pingcap/tidb/executor.(*HashJoinExec).Next.func2()
      /go/tidb/executor/join.go:685 +0x47
Goroutine 64 (running) created at:
  github.com/pingcap/tidb/executor.(*HashJoinExec).fetchAndBuildHashTable()
      /go/tidb/executor/join.go:728 +0x405
  github.com/pingcap/tidb/executor.(*HashJoinExec).Next.func1()
      /go/tidb/executor/join.go:687 +0xc8
  github.com/pingcap/tidb/util.WithRecovery()
      /go/tidb/util/misc.go:90 +0x6d
  github.com/pingcap/tidb/executor.(*HashJoinExec).Next.func2()
      /go/tidb/executor/join.go:685 +0x47
================== 
```

### PR DESCRIPTION
**Pull request by hawkingrei**
Signed-off-by: Weizhen Wang <wangweizhen@pingcap.com>

### What problem does this PR solve?

From the the stack of #34657. we can know ```hugeMemTableRetriever.dataForColumnsInTable``` can be executed in parallel. so they will always get schema and outname of view. so We can optimize it by run once.

Issue Number: close #34657

Problem Summary:

### What is changed and how it works?

use two maps to cache the result of schema view and outname view

### Check List

Tests 

- [x] Unit test
- [x] Integration test

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: executor/infoschema_reader.go [hunk 1]
```diff
@@ -801,27 +805,27 @@ ForColumnsTag:
...
-			charOctLen,           // CHARACTER_OCTET_LENGTH
-			numericPrecision,     // NUMERIC_PRECISION
-			numericScale,         // NUMERIC_SCALE
-			datetimePrecision,    // DATETIME_PRECISION
-			columnDesc.Charset,   // CHARACTER_SET_NAME
-			columnDesc.Collation, // COLLATION_NAME
-			columnType,           // COLUMN_TYPE
-			columnDesc.Key,       // COLUMN_KEY
-			columnDesc.Extra,     // EXTRA
+			infoschema.CatalogVal,                                                                // TABLE_CATALOG
```
Comment (hawkingrei) [inline executor/infoschema_reader.go:808]: just code alignment

**Review by tiancaiamao — APPROVED** [review]

**Review by qw4990 — APPROVED** [review]

**Pull request merged**